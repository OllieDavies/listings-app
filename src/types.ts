export type Holiday = {
  id: string
  hotelName: string
  location: string
  price: number
  overview: string
  departureAirport: string
  departureDate: string
  rating: number
  duration: string
  imageUrl: string
  party: Party
}

export type Party = {
  children?: number
  adults?: number
  infants?: number
}
