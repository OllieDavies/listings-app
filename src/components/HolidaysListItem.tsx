
import React, { useState } from "react";
import { Holiday, Party } from "../types";
import styles from "./HolidaysListItem.module.css"
import { StarRating } from "./StarRating";

interface Props {
    holiday: Holiday
}

const HolidaysListItem: React.FC<Props> = ({ holiday }) => {
    const [isOverviewVisible, setOverflowVisible] = useState<boolean>(false);

    const toggleOverflow = () => setOverflowVisible(!isOverviewVisible)

    return (
        <div className={styles.container}>
            <div className={styles.wrapper}>
                <div className={styles.imageContainer} style={{ backgroundImage: `url(${holiday.imageUrl})` }}>
                    <button onClick={toggleOverflow}>
                        <strong>Read {isOverviewVisible ? "less" : "more"}</strong> about this hotel
                    </button>
                </div>
                <div className={styles.textContainer}>
                    <h2 className={styles.textContainer__title}>{holiday.hotelName}</h2>
                    <p className={styles.textContainer__location}>{holiday.location}</p>
                    <StarRating rating={holiday.rating} />
                    <HolidayDetails {...holiday} />
                    <a className={styles.button} href={`./${holiday.id}`}>
                        Book now
                        <p className={styles.button__price}>Â£{holiday.price.toFixed(2)}</p>
                    </a>
                </div>
            </div>
            {isOverviewVisible === true &&
                <div className={`${styles.overviewContainer} ${styles.copy}`}>
                    <h4>Overview</h4>
                    <p>{holiday.overview}</p>
                </div>
            }
        </div>
    )
}

type DetailsProps = Pick<Holiday, "departureAirport" | "duration" | "departureDate" | "party">

const HolidayDetails: React.FC<DetailsProps> = (props) => (
    <div className={`${styles.copy} ${styles.textContainer__details}`}>
        <p><strong>{props.departureDate}</strong> for <strong>{props.duration}</strong></p>
        <p>{composePartyBreakdown(props.party)}</p>
        <p>departing from <strong>{props.departureAirport}</strong></p>
    </div>
)

const composePartyBreakdown = (party: Party) => {
    const entries = Object.entries(party)
    return entries
        .map<React.ReactNode>(item => (
            <React.Fragment key={item[0]}>
                <strong>{item[1]}</strong> {item[0]}
            </React.Fragment>
        ))
        .reduce((acc, current) => [acc, ", ", current])
}

export { HolidaysListItem, composePartyBreakdown }