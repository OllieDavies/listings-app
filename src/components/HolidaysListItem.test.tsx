import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { HolidaysListItem, composePartyBreakdown } from "./HolidaysListItem";
import { Holiday } from "../types";

const mockHoliday: Holiday = {
  id: "00001",
  hotelName: "Mock Hotel",
  location: "Princess Street, Manchester",
  price: 1,
  overview: "A mock overview",
  departureAirport: "Mock Airport",
  departureDate: "3rd July 2019",
  rating: 5,
  duration: "7 days",
  imageUrl: "/mock-image.png",
  party: {
    adults: 2
  }
}

test("renders the overview after Read More is clicked", () => {
  render(<HolidaysListItem holiday={mockHoliday} />);
  expect(screen.queryByText(/A mock overview/i)).not.toBeInTheDocument()
  const button = screen.getByText(/Read more/i);
  fireEvent.click(button);
  expect(screen.queryByText(/A mock overview/i)).toBeInTheDocument()
});

test("composePartyBreakdown matches snapshot", () => {
  expect(composePartyBreakdown({ adults: 2, children: 1 })).toMatchSnapshot();
})
