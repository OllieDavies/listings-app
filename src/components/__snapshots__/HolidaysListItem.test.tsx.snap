// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composePartyBreakdown matches snapshot 1`] = `
Array [
  <React.Fragment>
    <strong>
      2
    </strong>
     
    adults
  </React.Fragment>,
  ", ",
  <React.Fragment>
    <strong>
      1
    </strong>
     
    children
  </React.Fragment>,
]
`;
